python/cudf_polars/cudf_polars/utils/config.py

```diff
@@ -5,209 +5,304 @@
-        options = self.config_options
-        keys = name.split(".")
-        for k in keys[:-1]:
-            assert isinstance(options, dict)
-            options = options.get(k, {})
-        return options.get(keys[-1], default)
+    TASKS = "tasks"
+    RAPIDSMPF = "rapidsmpf"
 
-    def __hash__(self) -> int:
-        """Hash a ConfigOptions object."""
-        try:
-            return self._hash_value
-        except AttributeError:
-            self._hash_value = hash(json.dumps(self.config_options))
-            return self._hash_value
-
-    @staticmethod
-    def validate(config: dict) -> dict:
-        """
-        Validate a configuration-option dictionary.
 
-        Parameters
-        ----------
-        config
-            GPUEngine configuration options to validate.
+@dataclasses.dataclass(frozen=True)
+class ParquetOptions:
+    """
+    Configuration for the cudf-polars Parquet engine.
+
+    Parameters
+    ----------
+    chunked
+        Whether to use cudf's ``ChunkedParquetReader`` to read the parquet
+        dataset in chunks. This is useful when reading very large parquet
+        files.
+    chunk_read_limit
+        Limit on total number of bytes to be returned per read, or 0 if
+        there is no limit.
+    pass_read_limit
+        Limit on the amount of memory used for reading and decompressing data
+        or 0 if there is no limit.
+    """
 
-        Returns
-        -------
-        Valid config-option dictionary.
+    chunked: bool = True
+    chunk_read_limit: int = 0
+    pass_read_limit: int = 0
 
-        Raises
-        ------
-        ValueError
-            If the configuration contains unsupported options.
-        """
-        if unsupported := (
-            config.keys()
-            - {"raise_on_fail", "parquet_options", "executor", "executor_options"}
-        ):
+    def __post_init__(self) -> None:
+        if not isinstance(self.chunked, bool):
+            raise TypeError("chunked must be a bool")
+        if not isinstance(self.chunk_read_limit, int):
+            raise TypeError("chunk_read_limit must be an int")
+        if not isinstance(self.pass_read_limit, int):
+            raise TypeError("pass_read_limit must be an int")
+
+
+@dataclasses.dataclass(frozen=True, eq=True)
+class StreamingExecutor:
+    """
+    Configuration for the cudf-polars streaming executor.
+
+    Parameters
+    ----------
+    scheduler
+        The scheduler to use for the streaming executor. ``Scheduler.SYNCHRONOUS``
+        by default.
+
+        Note ``scheduler="distributed"`` requires a Dask cluster to be running.
+
+    fallback_mode
+        How to handle errors when the GPU engine fails to execute a query.
+        ``FallbackMode.WARN`` by default.
+    max_rows_per_partition
+        The maximum number of rows to process per partition. 1_000_000 by default.
+        When the number of rows exceeds this value, the query will be split into
+        multiple partitions and executed in parallel.
```

Comment:
*question*: Is this a hard limit or a guideline?

Reply:
This is only used for DataFrameScan, and is therefore mostly for testing/demonstration. I was thinking yesterday that we could try to replace `executor_options.partquet_blocksize` with `executor_options.blocksize` and use that number to determine the number of partitions instead. However, it's really nice to a have a specific row count for testing purposes.

Reply:
There shouldn't be any behavior change here, so it should be the same as on branch-25.06.  Looking at `lower_ir_node.register(DataFrameScan)` I see

```
    rows_per_partition = ir.config_options.executor.max_rows_per_partition

    nrows = max(ir.df.shape()[0], 1)
    count = math.ceil(nrows / rows_per_partition)

    if count > 1:
        ....
```

so it seems to be a hard limit there, and that's the only current usage.

However, I'm a little hesitant to document that it's a hard limit, in case we start using this value elsewhere and want some fuzziness. So I'd recommend either leaving it ambiguous or documenting it as a guideline. Do you have a preference?